// thirdweb/dist/cjs/extensions/modular/__generated__/ClaimableERC721/read/getSaleConfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isGetSaleConfigSupported = isGetSaleConfigSupported;
exports.decodeGetSaleConfigResult = decodeGetSaleConfigResult;
exports.getSaleConfig = getSaleConfig;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xcea943ee";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        name: "primarySaleRecipient",
        type: "address",
        internalType: "address",
    },
];
/**
 * Checks if the `getSaleConfig` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `getSaleConfig` method is supported.
 * @extension MODULAR
 * @example
 * ```ts
 * import { isGetSaleConfigSupported } from "thirdweb/extensions/modular";
 *
 * const supported = await isGetSaleConfigSupported(contract);
 * ```
 */
async function isGetSaleConfigSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Decodes the result of the getSaleConfig function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension MODULAR
 * @example
 * ```ts
 * import { decodeGetSaleConfigResult } from "thirdweb/extensions/modular";
 * const result = decodeGetSaleConfigResult("...");
 * ```
 */
function decodeGetSaleConfigResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "getSaleConfig" function on the contract.
 * @param options - The options for the getSaleConfig function.
 * @returns The parsed result of the function call.
 * @extension MODULAR
 * @example
 * ```ts
 * import { getSaleConfig } from "thirdweb/extensions/modular";
 *
 * const result = await getSaleConfig({
 *  contract,
 * });
 *
 * ```
 */
async function getSaleConfig(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
