// thirdweb/dist/cjs/extensions/modular/__generated__/MintableERC721/write/onUninstall.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isOnUninstallSupported = isOnUninstallSupported;
exports.encodeOnUninstallParams = encodeOnUninstallParams;
exports.encodeOnUninstall = encodeOnUninstall;
exports.onUninstall = onUninstall;
const prepare_contract_call_js_1 = require("../../../../../transaction/prepare-contract-call.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const once_js_1 = require("../../../../../utils/promise/once.js");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x8a91b0e3";
const FN_INPUTS = [
    {
        name: "data",
        type: "bytes",
        internalType: "bytes",
    },
];
const FN_OUTPUTS = [];
/**
 * Checks if the `onUninstall` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `onUninstall` method is supported.
 * @extension MODULAR
 * @example
 * ```ts
 * import { isOnUninstallSupported } from "thirdweb/extensions/modular";
 *
 * const supported = await isOnUninstallSupported(contract);
 * ```
 */
async function isOnUninstallSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "onUninstall" function.
 * @param options - The options for the onUninstall function.
 * @returns The encoded ABI parameters.
 * @extension MODULAR
 * @example
 * ```ts
 * import { encodeOnUninstallParams } "thirdweb/extensions/modular";
 * const result = encodeOnUninstallParams({
 *  data: ...,
 * });
 * ```
 */
function encodeOnUninstallParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.data]);
}
/**
 * Encodes the "onUninstall" function into a Hex string with its parameters.
 * @param options - The options for the onUninstall function.
 * @returns The encoded hexadecimal string.
 * @extension MODULAR
 * @example
 * ```ts
 * import { encodeOnUninstall } "thirdweb/extensions/modular";
 * const result = encodeOnUninstall({
 *  data: ...,
 * });
 * ```
 */
function encodeOnUninstall(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeOnUninstallParams(options).slice(2));
}
/**
 * Prepares a transaction to call the "onUninstall" function on the contract.
 * @param options - The options for the "onUninstall" function.
 * @returns A prepared transaction object.
 * @extension MODULAR
 * @example
 * ```ts
 * import { onUninstall } from "thirdweb/extensions/modular";
 *
 * const transaction = onUninstall({
 *  contract,
 *  data: ...,
 *  overrides: {
 *    ...
 *  }
 * });
 *
 * // Send the transaction
 * ...
 *
 * ```
 */
function onUninstall(options) {
    const asyncOptions = (0, once_js_1.once)(async () => {
        return "asyncParams" in options ? await options.asyncParams() : options;
    });
    return (0, prepare_contract_call_js_1.prepareContractCall)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: async () => {
            const resolvedOptions = await asyncOptions();
            return [resolvedOptions.data];
        },
        value: async () => (await asyncOptions()).overrides?.value,
        accessList: async () => (await asyncOptions()).overrides?.accessList,
        gas: async () => (await asyncOptions()).overrides?.gas,
        gasPrice: async () => (await asyncOptions()).overrides?.gasPrice,
        maxFeePerGas: async () => (await asyncOptions()).overrides?.maxFeePerGas,
        maxPriorityFeePerGas: async () => (await asyncOptions()).overrides?.maxPriorityFeePerGas,
        nonce: async () => (await asyncOptions()).overrides?.nonce,
        extraGas: async () => (await asyncOptions()).overrides?.extraGas,
    });
}
