// thirdweb/dist/cjs/extensions/modular/__generated__/MintableERC1155/write/beforeMintERC1155.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isBeforeMintERC1155Supported = isBeforeMintERC1155Supported;
exports.encodeBeforeMintERC1155Params = encodeBeforeMintERC1155Params;
exports.encodeBeforeMintERC1155 = encodeBeforeMintERC1155;
exports.beforeMintERC1155 = beforeMintERC1155;
const prepare_contract_call_js_1 = require("../../../../../transaction/prepare-contract-call.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const once_js_1 = require("../../../../../utils/promise/once.js");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x1e1dcb18";
const FN_INPUTS = [
    {
        name: "_to",
        type: "address",
        internalType: "address",
    },
    {
        name: "_id",
        type: "uint256",
        internalType: "uint256",
    },
    {
        name: "_quantity",
        type: "uint256",
        internalType: "uint256",
    },
    {
        name: "_data",
        type: "bytes",
        internalType: "bytes",
    },
];
const FN_OUTPUTS = [
    {
        name: "",
        type: "bytes",
        internalType: "bytes",
    },
];
/**
 * Checks if the `beforeMintERC1155` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `beforeMintERC1155` method is supported.
 * @extension MODULAR
 * @example
 * ```ts
 * import { isBeforeMintERC1155Supported } from "thirdweb/extensions/modular";
 *
 * const supported = await isBeforeMintERC1155Supported(contract);
 * ```
 */
async function isBeforeMintERC1155Supported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "beforeMintERC1155" function.
 * @param options - The options for the beforeMintERC1155 function.
 * @returns The encoded ABI parameters.
 * @extension MODULAR
 * @example
 * ```ts
 * import { encodeBeforeMintERC1155Params } "thirdweb/extensions/modular";
 * const result = encodeBeforeMintERC1155Params({
 *  to: ...,
 *  id: ...,
 *  quantity: ...,
 *  data: ...,
 * });
 * ```
 */
function encodeBeforeMintERC1155Params(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [
        options.to,
        options.id,
        options.quantity,
        options.data,
    ]);
}
/**
 * Encodes the "beforeMintERC1155" function into a Hex string with its parameters.
 * @param options - The options for the beforeMintERC1155 function.
 * @returns The encoded hexadecimal string.
 * @extension MODULAR
 * @example
 * ```ts
 * import { encodeBeforeMintERC1155 } "thirdweb/extensions/modular";
 * const result = encodeBeforeMintERC1155({
 *  to: ...,
 *  id: ...,
 *  quantity: ...,
 *  data: ...,
 * });
 * ```
 */
function encodeBeforeMintERC1155(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeBeforeMintERC1155Params(options).slice(2));
}
/**
 * Prepares a transaction to call the "beforeMintERC1155" function on the contract.
 * @param options - The options for the "beforeMintERC1155" function.
 * @returns A prepared transaction object.
 * @extension MODULAR
 * @example
 * ```ts
 * import { beforeMintERC1155 } from "thirdweb/extensions/modular";
 *
 * const transaction = beforeMintERC1155({
 *  contract,
 *  to: ...,
 *  id: ...,
 *  quantity: ...,
 *  data: ...,
 *  overrides: {
 *    ...
 *  }
 * });
 *
 * // Send the transaction
 * ...
 *
 * ```
 */
function beforeMintERC1155(options) {
    const asyncOptions = (0, once_js_1.once)(async () => {
        return "asyncParams" in options ? await options.asyncParams() : options;
    });
    return (0, prepare_contract_call_js_1.prepareContractCall)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: async () => {
            const resolvedOptions = await asyncOptions();
            return [
                resolvedOptions.to,
                resolvedOptions.id,
                resolvedOptions.quantity,
                resolvedOptions.data,
            ];
        },
        value: async () => (await asyncOptions()).overrides?.value,
        accessList: async () => (await asyncOptions()).overrides?.accessList,
        gas: async () => (await asyncOptions()).overrides?.gas,
        gasPrice: async () => (await asyncOptions()).overrides?.gasPrice,
        maxFeePerGas: async () => (await asyncOptions()).overrides?.maxFeePerGas,
        maxPriorityFeePerGas: async () => (await asyncOptions()).overrides?.maxPriorityFeePerGas,
        nonce: async () => (await asyncOptions()).overrides?.nonce,
        extraGas: async () => (await asyncOptions()).overrides?.extraGas,
    });
}
