// thirdweb/dist/esm/extensions/modular/__generated__/ClaimableERC20/read/encodeBytesOnInstall.js
import { readContract } from "../../../../../transaction/read-contract.js";
import { encodeAbiParameters } from "../../../../../utils/abi/encodeAbiParameters.js";
import { decodeAbiParameters } from "viem";
import { detectMethod } from "../../../../../utils/bytecode/detectExtension.js";
export const FN_SELECTOR = "0x5d4c0b89";
const FN_INPUTS = [
    {
        name: "primarySaleRecipient",
        type: "address",
        internalType: "address",
    },
];
const FN_OUTPUTS = [
    {
        name: "",
        type: "bytes",
        internalType: "bytes",
    },
];
/**
 * Checks if the `encodeBytesOnInstall` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `encodeBytesOnInstall` method is supported.
 * @extension MODULAR
 * @example
 * ```ts
 * import { isEncodeBytesOnInstallSupported } from "thirdweb/extensions/modular";
 *
 * const supported = await isEncodeBytesOnInstallSupported(contract);
 * ```
 */
export async function isEncodeBytesOnInstallSupported(contract) {
    return detectMethod({
        contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "encodeBytesOnInstall" function.
 * @param options - The options for the encodeBytesOnInstall function.
 * @returns The encoded ABI parameters.
 * @extension MODULAR
 * @example
 * ```ts
 * import { encodeEncodeBytesOnInstallParams } "thirdweb/extensions/modular";
 * const result = encodeEncodeBytesOnInstallParams({
 *  primarySaleRecipient: ...,
 * });
 * ```
 */
export function encodeEncodeBytesOnInstallParams(options) {
    return encodeAbiParameters(FN_INPUTS, [options.primarySaleRecipient]);
}
/**
 * Encodes the "encodeBytesOnInstall" function into a Hex string with its parameters.
 * @param options - The options for the encodeBytesOnInstall function.
 * @returns The encoded hexadecimal string.
 * @extension MODULAR
 * @example
 * ```ts
 * import { encodeEncodeBytesOnInstall } "thirdweb/extensions/modular";
 * const result = encodeEncodeBytesOnInstall({
 *  primarySaleRecipient: ...,
 * });
 * ```
 */
export function encodeEncodeBytesOnInstall(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (FN_SELECTOR +
        encodeEncodeBytesOnInstallParams(options).slice(2));
}
/**
 * Decodes the result of the encodeBytesOnInstall function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension MODULAR
 * @example
 * ```ts
 * import { decodeEncodeBytesOnInstallResult } from "thirdweb/extensions/modular";
 * const result = decodeEncodeBytesOnInstallResult("...");
 * ```
 */
export function decodeEncodeBytesOnInstallResult(result) {
    return decodeAbiParameters(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "encodeBytesOnInstall" function on the contract.
 * @param options - The options for the encodeBytesOnInstall function.
 * @returns The parsed result of the function call.
 * @extension MODULAR
 * @example
 * ```ts
 * import { encodeBytesOnInstall } from "thirdweb/extensions/modular";
 *
 * const result = await encodeBytesOnInstall({
 *  contract,
 *  primarySaleRecipient: ...,
 * });
 *
 * ```
 */
export async function encodeBytesOnInstall(options) {
    return readContract({
        contract: options.contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.primarySaleRecipient],
    });
}
