// thirdweb/dist/esm/extensions/modular/__generated__/ClaimableERC20/read/getSaleConfig.js
import { readContract } from "../../../../../transaction/read-contract.js";
import { decodeAbiParameters } from "viem";
import { detectMethod } from "../../../../../utils/bytecode/detectExtension.js";
export const FN_SELECTOR = "0xcea943ee";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        name: "primarySaleRecipient",
        type: "address",
        internalType: "address",
    },
];
/**
 * Checks if the `getSaleConfig` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `getSaleConfig` method is supported.
 * @extension MODULAR
 * @example
 * ```ts
 * import { isGetSaleConfigSupported } from "thirdweb/extensions/modular";
 *
 * const supported = await isGetSaleConfigSupported(contract);
 * ```
 */
export async function isGetSaleConfigSupported(contract) {
    return detectMethod({
        contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Decodes the result of the getSaleConfig function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension MODULAR
 * @example
 * ```ts
 * import { decodeGetSaleConfigResult } from "thirdweb/extensions/modular";
 * const result = decodeGetSaleConfigResult("...");
 * ```
 */
export function decodeGetSaleConfigResult(result) {
    return decodeAbiParameters(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "getSaleConfig" function on the contract.
 * @param options - The options for the getSaleConfig function.
 * @returns The parsed result of the function call.
 * @extension MODULAR
 * @example
 * ```ts
 * import { getSaleConfig } from "thirdweb/extensions/modular";
 *
 * const result = await getSaleConfig({
 *  contract,
 * });
 *
 * ```
 */
export async function getSaleConfig(options) {
    return readContract({
        contract: options.contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
