// thirdweb/dist/cjs/extensions/lens/__generated__/LensHandle/read/getHandle.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isGetHandleSupported = isGetHandleSupported;
exports.encodeGetHandleParams = encodeGetHandleParams;
exports.encodeGetHandle = encodeGetHandle;
exports.decodeGetHandleResult = decodeGetHandleResult;
exports.getHandle = getHandle;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xec81d194";
const FN_INPUTS = [
    {
        type: "uint256",
        name: "tokenId",
    },
];
const FN_OUTPUTS = [
    {
        type: "string",
    },
];
/**
 * Checks if the `getHandle` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `getHandle` method is supported.
 * @extension LENS
 * @example
 * ```ts
 * import { isGetHandleSupported } from "thirdweb/extensions/lens";
 *
 * const supported = await isGetHandleSupported(contract);
 * ```
 */
async function isGetHandleSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "getHandle" function.
 * @param options - The options for the getHandle function.
 * @returns The encoded ABI parameters.
 * @extension LENS
 * @example
 * ```ts
 * import { encodeGetHandleParams } "thirdweb/extensions/lens";
 * const result = encodeGetHandleParams({
 *  tokenId: ...,
 * });
 * ```
 */
function encodeGetHandleParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.tokenId]);
}
/**
 * Encodes the "getHandle" function into a Hex string with its parameters.
 * @param options - The options for the getHandle function.
 * @returns The encoded hexadecimal string.
 * @extension LENS
 * @example
 * ```ts
 * import { encodeGetHandle } "thirdweb/extensions/lens";
 * const result = encodeGetHandle({
 *  tokenId: ...,
 * });
 * ```
 */
function encodeGetHandle(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeGetHandleParams(options).slice(2));
}
/**
 * Decodes the result of the getHandle function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension LENS
 * @example
 * ```ts
 * import { decodeGetHandleResult } from "thirdweb/extensions/lens";
 * const result = decodeGetHandleResult("...");
 * ```
 */
function decodeGetHandleResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "getHandle" function on the contract.
 * @param options - The options for the getHandle function.
 * @returns The parsed result of the function call.
 * @extension LENS
 * @example
 * ```ts
 * import { getHandle } from "thirdweb/extensions/lens";
 *
 * const result = await getHandle({
 *  contract,
 *  tokenId: ...,
 * });
 *
 * ```
 */
async function getHandle(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.tokenId],
    });
}
