// thirdweb/dist/cjs/extensions/lens/__generated__/LensHandle/read/getTokenId.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isGetTokenIdSupported = isGetTokenIdSupported;
exports.encodeGetTokenIdParams = encodeGetTokenIdParams;
exports.encodeGetTokenId = encodeGetTokenId;
exports.decodeGetTokenIdResult = decodeGetTokenIdResult;
exports.getTokenId = getTokenId;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x1e7663bc";
const FN_INPUTS = [
    {
        type: "string",
        name: "localName",
    },
];
const FN_OUTPUTS = [
    {
        type: "uint256",
    },
];
/**
 * Checks if the `getTokenId` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `getTokenId` method is supported.
 * @extension LENS
 * @example
 * ```ts
 * import { isGetTokenIdSupported } from "thirdweb/extensions/lens";
 *
 * const supported = await isGetTokenIdSupported(contract);
 * ```
 */
async function isGetTokenIdSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "getTokenId" function.
 * @param options - The options for the getTokenId function.
 * @returns The encoded ABI parameters.
 * @extension LENS
 * @example
 * ```ts
 * import { encodeGetTokenIdParams } "thirdweb/extensions/lens";
 * const result = encodeGetTokenIdParams({
 *  localName: ...,
 * });
 * ```
 */
function encodeGetTokenIdParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.localName]);
}
/**
 * Encodes the "getTokenId" function into a Hex string with its parameters.
 * @param options - The options for the getTokenId function.
 * @returns The encoded hexadecimal string.
 * @extension LENS
 * @example
 * ```ts
 * import { encodeGetTokenId } "thirdweb/extensions/lens";
 * const result = encodeGetTokenId({
 *  localName: ...,
 * });
 * ```
 */
function encodeGetTokenId(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeGetTokenIdParams(options).slice(2));
}
/**
 * Decodes the result of the getTokenId function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension LENS
 * @example
 * ```ts
 * import { decodeGetTokenIdResult } from "thirdweb/extensions/lens";
 * const result = decodeGetTokenIdResult("...");
 * ```
 */
function decodeGetTokenIdResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "getTokenId" function on the contract.
 * @param options - The options for the getTokenId function.
 * @returns The parsed result of the function call.
 * @extension LENS
 * @example
 * ```ts
 * import { getTokenId } from "thirdweb/extensions/lens";
 *
 * const result = await getTokenId({
 *  contract,
 *  localName: ...,
 * });
 *
 * ```
 */
async function getTokenId(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.localName],
    });
}
