// thirdweb/dist/esm/extensions/modular/__generated__/MintableERC721/write/beforeMintERC721.js
import { prepareContractCall } from "../../../../../transaction/prepare-contract-call.js";
import { encodeAbiParameters } from "../../../../../utils/abi/encodeAbiParameters.js";
import { once } from "../../../../../utils/promise/once.js";
import { detectMethod } from "../../../../../utils/bytecode/detectExtension.js";
export const FN_SELECTOR = "0x765e8093";
const FN_INPUTS = [
    {
        name: "_to",
        type: "address",
        internalType: "address",
    },
    {
        name: "_startTokenId",
        type: "uint256",
        internalType: "uint256",
    },
    {
        name: "_quantity",
        type: "uint256",
        internalType: "uint256",
    },
    {
        name: "_data",
        type: "bytes",
        internalType: "bytes",
    },
];
const FN_OUTPUTS = [
    {
        name: "",
        type: "bytes",
        internalType: "bytes",
    },
];
/**
 * Checks if the `beforeMintERC721` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `beforeMintERC721` method is supported.
 * @extension MODULAR
 * @example
 * ```ts
 * import { isBeforeMintERC721Supported } from "thirdweb/extensions/modular";
 *
 * const supported = await isBeforeMintERC721Supported(contract);
 * ```
 */
export async function isBeforeMintERC721Supported(contract) {
    return detectMethod({
        contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "beforeMintERC721" function.
 * @param options - The options for the beforeMintERC721 function.
 * @returns The encoded ABI parameters.
 * @extension MODULAR
 * @example
 * ```ts
 * import { encodeBeforeMintERC721Params } "thirdweb/extensions/modular";
 * const result = encodeBeforeMintERC721Params({
 *  to: ...,
 *  startTokenId: ...,
 *  quantity: ...,
 *  data: ...,
 * });
 * ```
 */
export function encodeBeforeMintERC721Params(options) {
    return encodeAbiParameters(FN_INPUTS, [
        options.to,
        options.startTokenId,
        options.quantity,
        options.data,
    ]);
}
/**
 * Encodes the "beforeMintERC721" function into a Hex string with its parameters.
 * @param options - The options for the beforeMintERC721 function.
 * @returns The encoded hexadecimal string.
 * @extension MODULAR
 * @example
 * ```ts
 * import { encodeBeforeMintERC721 } "thirdweb/extensions/modular";
 * const result = encodeBeforeMintERC721({
 *  to: ...,
 *  startTokenId: ...,
 *  quantity: ...,
 *  data: ...,
 * });
 * ```
 */
export function encodeBeforeMintERC721(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (FN_SELECTOR +
        encodeBeforeMintERC721Params(options).slice(2));
}
/**
 * Prepares a transaction to call the "beforeMintERC721" function on the contract.
 * @param options - The options for the "beforeMintERC721" function.
 * @returns A prepared transaction object.
 * @extension MODULAR
 * @example
 * ```ts
 * import { beforeMintERC721 } from "thirdweb/extensions/modular";
 *
 * const transaction = beforeMintERC721({
 *  contract,
 *  to: ...,
 *  startTokenId: ...,
 *  quantity: ...,
 *  data: ...,
 *  overrides: {
 *    ...
 *  }
 * });
 *
 * // Send the transaction
 * ...
 *
 * ```
 */
export function beforeMintERC721(options) {
    const asyncOptions = once(async () => {
        return "asyncParams" in options ? await options.asyncParams() : options;
    });
    return prepareContractCall({
        contract: options.contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: async () => {
            const resolvedOptions = await asyncOptions();
            return [
                resolvedOptions.to,
                resolvedOptions.startTokenId,
                resolvedOptions.quantity,
                resolvedOptions.data,
            ];
        },
        value: async () => (await asyncOptions()).overrides?.value,
        accessList: async () => (await asyncOptions()).overrides?.accessList,
        gas: async () => (await asyncOptions()).overrides?.gas,
        gasPrice: async () => (await asyncOptions()).overrides?.gasPrice,
        maxFeePerGas: async () => (await asyncOptions()).overrides?.maxFeePerGas,
        maxPriorityFeePerGas: async () => (await asyncOptions()).overrides?.maxPriorityFeePerGas,
        nonce: async () => (await asyncOptions()).overrides?.nonce,
        extraGas: async () => (await asyncOptions()).overrides?.extraGas,
    });
}
