// thirdweb/dist/esm/wallets/coinbase/coinbaseMobileSDK.js
import { configure, handleResponse } from "@coinbase/wallet-mobile-sdk";
import { Linking } from "react-native";
let _provider;
/**
 * @internal
 */
export async function getCoinbaseMobileProvider(options) {
    if (!_provider) {
        const callbackURL = new URL(options?.mobileConfig?.callbackURL || "https://thirdweb.com/wsegue");
        const mobileProvider = (await initMobileProvider({
            chain: options?.chains ? options.chains[0] : undefined,
            ...options?.mobileConfig,
        }));
        _provider = mobileProvider;
        Linking.addEventListener("url", ({ url }) => {
            const incomingUrl = new URL(url);
            if (callbackURL &&
                incomingUrl.host === callbackURL.host &&
                incomingUrl.protocol === callbackURL.protocol &&
                incomingUrl.hostname === callbackURL.hostname) {
                // @ts-expect-error - Passing a URL object to handleResponse crashes the function
                handleResponse(url);
            }
        });
        return mobileProvider;
    }
    return _provider;
}
async function initMobileProvider(args) {
    configure({
        callbackURL: new URL(args.callbackURL || "https://thirdweb.com/wsegue"),
        hostURL: new URL(args.hostURL || "https://wallet.coinbase.com/wsegue"),
        hostPackageName: args.hostPackageName || "org.toshi",
    });
    let CoinbaseWalletMobileSDK = (await import(
    // @ts-ignore this is only visible from RN
    "@coinbase/wallet-mobile-sdk/build/WalletMobileSDKEVMProvider")).WalletMobileSDKEVMProvider;
    if (typeof CoinbaseWalletMobileSDK !== "function" &&
        // @ts-ignore this is only visible from RN
        typeof CoinbaseWalletMobileSDK.default === "function") {
        CoinbaseWalletMobileSDK = CoinbaseWalletMobileSDK.default;
    }
    return new CoinbaseWalletMobileSDK({
        jsonRpcUrl: args.chain?.rpc,
        chainId: args.chain?.id,
    });
}
