// thirdweb/dist/esm/extensions/modular/__generated__/BatchMetadataERC721/read/getAllMetadataBatches.js
import { readContract } from "../../../../../transaction/read-contract.js";
import { decodeAbiParameters } from "viem";
import { detectMethod } from "../../../../../utils/bytecode/detectExtension.js";
export const FN_SELECTOR = "0xe6c23512";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        name: "",
        type: "tuple[]",
        internalType: "struct BatchMetadataERC721.MetadataBatch[]",
        components: [
            {
                name: "startTokenIdInclusive",
                type: "uint256",
                internalType: "uint256",
            },
            {
                name: "endTokenIdInclusive",
                type: "uint256",
                internalType: "uint256",
            },
            {
                name: "baseURI",
                type: "string",
                internalType: "string",
            },
        ],
    },
];
/**
 * Checks if the `getAllMetadataBatches` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `getAllMetadataBatches` method is supported.
 * @extension MODULAR
 * @example
 * ```ts
 * import { isGetAllMetadataBatchesSupported } from "thirdweb/extensions/modular";
 *
 * const supported = await isGetAllMetadataBatchesSupported(contract);
 * ```
 */
export async function isGetAllMetadataBatchesSupported(contract) {
    return detectMethod({
        contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Decodes the result of the getAllMetadataBatches function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension MODULAR
 * @example
 * ```ts
 * import { decodeGetAllMetadataBatchesResult } from "thirdweb/extensions/modular";
 * const result = decodeGetAllMetadataBatchesResult("...");
 * ```
 */
export function decodeGetAllMetadataBatchesResult(result) {
    return decodeAbiParameters(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "getAllMetadataBatches" function on the contract.
 * @param options - The options for the getAllMetadataBatches function.
 * @returns The parsed result of the function call.
 * @extension MODULAR
 * @example
 * ```ts
 * import { getAllMetadataBatches } from "thirdweb/extensions/modular";
 *
 * const result = await getAllMetadataBatches({
 *  contract,
 * });
 *
 * ```
 */
export async function getAllMetadataBatches(options) {
    return readContract({
        contract: options.contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
