// thirdweb/dist/cjs/extensions/modular/__generated__/MintableERC20/read/encodeBytesBeforeMintERC20.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isEncodeBytesBeforeMintERC20Supported = isEncodeBytesBeforeMintERC20Supported;
exports.encodeEncodeBytesBeforeMintERC20Params = encodeEncodeBytesBeforeMintERC20Params;
exports.encodeEncodeBytesBeforeMintERC20 = encodeEncodeBytesBeforeMintERC20;
exports.decodeEncodeBytesBeforeMintERC20Result = decodeEncodeBytesBeforeMintERC20Result;
exports.encodeBytesBeforeMintERC20 = encodeBytesBeforeMintERC20;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xc7a090f4";
const FN_INPUTS = [
    {
        name: "params",
        type: "tuple",
        internalType: "struct MintableERC20.MintParamsERC20",
        components: [
            {
                name: "request",
                type: "tuple",
                internalType: "struct MintableERC20.MintRequestERC20",
                components: [
                    {
                        name: "startTimestamp",
                        type: "uint48",
                        internalType: "uint48",
                    },
                    {
                        name: "endTimestamp",
                        type: "uint48",
                        internalType: "uint48",
                    },
                    {
                        name: "recipient",
                        type: "address",
                        internalType: "address",
                    },
                    {
                        name: "quantity",
                        type: "uint256",
                        internalType: "uint256",
                    },
                    {
                        name: "currency",
                        type: "address",
                        internalType: "address",
                    },
                    {
                        name: "pricePerUnit",
                        type: "uint256",
                        internalType: "uint256",
                    },
                    {
                        name: "uid",
                        type: "bytes32",
                        internalType: "bytes32",
                    },
                ],
            },
            {
                name: "signature",
                type: "bytes",
                internalType: "bytes",
            },
        ],
    },
];
const FN_OUTPUTS = [
    {
        name: "",
        type: "bytes",
        internalType: "bytes",
    },
];
/**
 * Checks if the `encodeBytesBeforeMintERC20` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `encodeBytesBeforeMintERC20` method is supported.
 * @extension MODULAR
 * @example
 * ```ts
 * import { isEncodeBytesBeforeMintERC20Supported } from "thirdweb/extensions/modular";
 *
 * const supported = await isEncodeBytesBeforeMintERC20Supported(contract);
 * ```
 */
async function isEncodeBytesBeforeMintERC20Supported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "encodeBytesBeforeMintERC20" function.
 * @param options - The options for the encodeBytesBeforeMintERC20 function.
 * @returns The encoded ABI parameters.
 * @extension MODULAR
 * @example
 * ```ts
 * import { encodeEncodeBytesBeforeMintERC20Params } "thirdweb/extensions/modular";
 * const result = encodeEncodeBytesBeforeMintERC20Params({
 *  params: ...,
 * });
 * ```
 */
function encodeEncodeBytesBeforeMintERC20Params(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.params]);
}
/**
 * Encodes the "encodeBytesBeforeMintERC20" function into a Hex string with its parameters.
 * @param options - The options for the encodeBytesBeforeMintERC20 function.
 * @returns The encoded hexadecimal string.
 * @extension MODULAR
 * @example
 * ```ts
 * import { encodeEncodeBytesBeforeMintERC20 } "thirdweb/extensions/modular";
 * const result = encodeEncodeBytesBeforeMintERC20({
 *  params: ...,
 * });
 * ```
 */
function encodeEncodeBytesBeforeMintERC20(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeEncodeBytesBeforeMintERC20Params(options).slice(2));
}
/**
 * Decodes the result of the encodeBytesBeforeMintERC20 function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension MODULAR
 * @example
 * ```ts
 * import { decodeEncodeBytesBeforeMintERC20Result } from "thirdweb/extensions/modular";
 * const result = decodeEncodeBytesBeforeMintERC20Result("...");
 * ```
 */
function decodeEncodeBytesBeforeMintERC20Result(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "encodeBytesBeforeMintERC20" function on the contract.
 * @param options - The options for the encodeBytesBeforeMintERC20 function.
 * @returns The parsed result of the function call.
 * @extension MODULAR
 * @example
 * ```ts
 * import { encodeBytesBeforeMintERC20 } from "thirdweb/extensions/modular";
 *
 * const result = await encodeBytesBeforeMintERC20({
 *  contract,
 *  params: ...,
 * });
 *
 * ```
 */
async function encodeBytesBeforeMintERC20(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.params],
    });
}
