// thirdweb/dist/cjs/extensions/modular/__generated__/BatchMetadataERC1155/read/getAllMetadataBatches.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isGetAllMetadataBatchesSupported = isGetAllMetadataBatchesSupported;
exports.decodeGetAllMetadataBatchesResult = decodeGetAllMetadataBatchesResult;
exports.getAllMetadataBatches = getAllMetadataBatches;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xe6c23512";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        name: "",
        type: "tuple[]",
        internalType: "struct BatchMetadataERC721.MetadataBatch[]",
        components: [
            {
                name: "startTokenIdInclusive",
                type: "uint256",
                internalType: "uint256",
            },
            {
                name: "endTokenIdInclusive",
                type: "uint256",
                internalType: "uint256",
            },
            {
                name: "baseURI",
                type: "string",
                internalType: "string",
            },
        ],
    },
];
/**
 * Checks if the `getAllMetadataBatches` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `getAllMetadataBatches` method is supported.
 * @extension MODULAR
 * @example
 * ```ts
 * import { isGetAllMetadataBatchesSupported } from "thirdweb/extensions/modular";
 *
 * const supported = await isGetAllMetadataBatchesSupported(contract);
 * ```
 */
async function isGetAllMetadataBatchesSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Decodes the result of the getAllMetadataBatches function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension MODULAR
 * @example
 * ```ts
 * import { decodeGetAllMetadataBatchesResult } from "thirdweb/extensions/modular";
 * const result = decodeGetAllMetadataBatchesResult("...");
 * ```
 */
function decodeGetAllMetadataBatchesResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "getAllMetadataBatches" function on the contract.
 * @param options - The options for the getAllMetadataBatches function.
 * @returns The parsed result of the function call.
 * @extension MODULAR
 * @example
 * ```ts
 * import { getAllMetadataBatches } from "thirdweb/extensions/modular";
 *
 * const result = await getAllMetadataBatches({
 *  contract,
 * });
 *
 * ```
 */
async function getAllMetadataBatches(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
