// thirdweb/dist/cjs/react/web/wallets/shared/ConnectWalletSocialOptions.js
"use strict";
"use client";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConnectWalletSocialOptions = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const styled_1 = require("@emotion/styled");
const react_query_1 = require("@tanstack/react-query");
const react_1 = require("react");
const webStorage_js_1 = require("../../../../utils/storage/webStorage.js");
const get_ecosystem_wallet_auth_options_js_1 = require("../../../../wallets/ecosystem/get-ecosystem-wallet-auth-options.js");
const is_ecosystem_wallet_js_1 = require("../../../../wallets/ecosystem/is-ecosystem-wallet.js");
const oauth_js_1 = require("../../../../wallets/in-app/web/lib/auth/oauth.js");
const types_js_1 = require("../../../../wallets/types.js");
const CustomThemeProvider_js_1 = require("../../../core/design-system/CustomThemeProvider.js");
const index_js_1 = require("../../../core/design-system/index.js");
const socialIcons_js_1 = require("../../../core/utils/socialIcons.js");
const storage_js_1 = require("../../../core/utils/storage.js");
const wallet_ui_states_provider_js_1 = require("../../providers/wallet-ui-states-provider.js");
const WalletTypeRowButton_js_1 = require("../../ui/ConnectWallet/WalletTypeRowButton.js");
const Img_js_1 = require("../../ui/components/Img.js");
const TextDivider_js_1 = require("../../ui/components/TextDivider.js");
const basic_js_1 = require("../../ui/components/basic.js");
const buttons_js_1 = require("../../ui/components/buttons.js");
const InputSelectionUI_js_1 = require("../in-app/InputSelectionUI.js");
const validateEmail_js_1 = require("../in-app/validateEmail.js");
const LoadingScreen_js_1 = require("./LoadingScreen.js");
const oauthSignIn_js_1 = require("./oauthSignIn.js");
const defaultAuthOptions = [
    "email",
    "phone",
    "google",
    "apple",
    "facebook",
    "passkey",
];
/**
 * @internal
 */
const ConnectWalletSocialOptions = (props) => {
    const locale = props.locale;
    const { wallet } = props;
    const setData = (0, wallet_ui_states_provider_js_1.useSetSelectionData)();
    const themeObj = (0, CustomThemeProvider_js_1.useCustomTheme)();
    const loginMethodsLabel = {
        google: locale.signInWithGoogle,
        facebook: locale.signInWithFacebook,
        apple: locale.signInWithApple,
        discord: locale.signInWithDiscord,
        farcaster: "Farcaster",
        telegram: "Telegram",
    };
    const { data: ecosystemAuthOptions, isLoading } = (0, react_query_1.useQuery)({
        queryKey: ["auth-options", wallet.id],
        queryFn: async () => {
            if ((0, is_ecosystem_wallet_js_1.isEcosystemWallet)(wallet)) {
                return (0, get_ecosystem_wallet_auth_options_js_1.getEcosystemWalletAuthOptions)(wallet.id);
            }
            return null;
        },
        enabled: (0, is_ecosystem_wallet_js_1.isEcosystemWallet)(wallet),
        retry: false,
    });
    const authOptions = (0, is_ecosystem_wallet_js_1.isEcosystemWallet)(wallet)
        ? ecosystemAuthOptions ?? defaultAuthOptions
        : wallet.getConfig()?.auth?.options ?? defaultAuthOptions;
    const emailIndex = authOptions.indexOf("email");
    const isEmailEnabled = emailIndex !== -1;
    const phoneIndex = authOptions.indexOf("phone");
    const isPhoneEnabled = phoneIndex !== -1;
    const [manualInputMode, setManualInputMode] = (0, react_1.useState)(null);
    const inputMode = (0, react_1.useMemo)(() => {
        if (manualInputMode) {
            return manualInputMode;
        }
        if (isEmailEnabled && isPhoneEnabled) {
            return emailIndex < phoneIndex ? "email" : "phone";
        }
        if (isEmailEnabled) {
            return "email";
        }
        if (isPhoneEnabled) {
            return "phone";
        }
        return "none";
    }, [isEmailEnabled, isPhoneEnabled, emailIndex, phoneIndex, manualInputMode]);
    if ((0, is_ecosystem_wallet_js_1.isEcosystemWallet)(wallet) && isLoading) {
        return (0, jsx_runtime_1.jsx)(LoadingScreen_js_1.LoadingScreen, {});
    }
    const passKeyEnabled = authOptions.includes("passkey");
    const placeholder = inputMode === "email" ? locale.emailPlaceholder : locale.phonePlaceholder;
    const emptyErrorMessage = inputMode === "email" ? locale.emailRequired : locale.phoneRequired;
    let type = "text";
    if (inputMode === "email") {
        type = "email";
    }
    else if (inputMode === "phone") {
        type = "tel";
    }
    const socialLogins = authOptions.filter((o) => types_js_1.socialAuthOptions.includes(o));
    const hasSocialLogins = socialLogins.length > 0;
    const ecosystemInfo = (0, is_ecosystem_wallet_js_1.isEcosystemWallet)(wallet)
        ? {
            id: wallet.id,
            partnerId: wallet.getConfig()?.partnerId,
        }
        : undefined;
    // Need to trigger login on button click to avoid popup from being blocked
    const handleSocialLogin = async (strategy) => {
        const walletConfig = wallet.getConfig();
        if (walletConfig &&
            "auth" in walletConfig &&
            walletConfig?.auth?.mode === "redirect") {
            return (0, oauth_js_1.loginWithOauthRedirect)({
                authOption: strategy,
                client: props.client,
                ecosystem: ecosystemInfo,
            });
        }
        try {
            const socialLoginWindow = (0, oauthSignIn_js_1.openOauthSignInWindow)({
                authOption: strategy,
                themeObj,
                client: props.client,
                ecosystem: ecosystemInfo,
            });
            if (!socialLoginWindow) {
                throw new Error("Failed to open login window");
            }
            const connectOptions = {
                chain: props.chain,
                client: props.client,
                strategy,
                openedWindow: socialLoginWindow,
                closeOpenedWindow: (openedWindow) => {
                    openedWindow.close();
                },
            };
            const connectPromise = (0, is_ecosystem_wallet_js_1.isEcosystemWallet)(wallet)
                ? wallet.connect({
                    ...connectOptions,
                    ecosystem: {
                        id: wallet.id,
                        partnerId: wallet.getConfig()?.partnerId,
                    },
                })
                : wallet.connect(connectOptions);
            await (0, storage_js_1.setLastAuthProvider)(strategy, webStorage_js_1.webLocalStorage);
            setData({
                socialLogin: {
                    type: strategy,
                    connectionPromise: connectPromise,
                },
            });
            props.select(); // show Connect UI
            // Note: do not call done() here, it will be called SocialLogin component
            // we simply trigger the connect and save promise here - its resolution is handled in SocialLogin
        }
        catch (e) {
            console.error(`Error signing in with ${strategy}`, e);
        }
    };
    function handlePassKeyLogin() {
        setData({
            passkeyLogin: true,
        });
        props.select();
    }
    const showOnlyIcons = socialLogins.length > 2;
    return ((0, jsx_runtime_1.jsxs)(basic_js_1.Container, { flex: "column", gap: "md", style: {
            position: "relative",
        }, children: [hasSocialLogins && ((0, jsx_runtime_1.jsx)(basic_js_1.Container, { flex: "row", center: "x", gap: socialLogins.length > 4 ? "xs" : "sm", style: {
                    justifyContent: "space-between",
                    display: "grid",
                    gridTemplateColumns: `repeat(${socialLogins.length}, 1fr)`,
                }, children: socialLogins.map((loginMethod) => {
                    const imgIconSize = (() => {
                        if (!showOnlyIcons) {
                            return index_js_1.iconSize.md;
                        }
                        else {
                            if (socialLogins.length > 4) {
                                return index_js_1.iconSize.md;
                            }
                            return index_js_1.iconSize.lg;
                        }
                    })();
                    return ((0, jsx_runtime_1.jsxs)(SocialButton, { "aria-label": `Login with ${loginMethod}`, "data-variant": showOnlyIcons ? "icon" : "full", variant: "outline", onClick: () => {
                            handleSocialLogin(loginMethod);
                        }, children: [(0, jsx_runtime_1.jsx)(Img_js_1.Img, { src: socialIcons_js_1.socialIcons[loginMethod], width: imgIconSize, height: imgIconSize, client: props.client }), !showOnlyIcons &&
                                `${socialLogins.length === 1 ? "Continue with " : ""}${loginMethodsLabel[loginMethod]}`] }, loginMethod));
                }) })), props.size === "wide" &&
                hasSocialLogins &&
                (isEmailEnabled || isPhoneEnabled) && (0, jsx_runtime_1.jsx)(TextDivider_js_1.TextDivider, { text: locale.or }), isEmailEnabled && ((0, jsx_runtime_1.jsx)(jsx_runtime_1.Fragment, { children: inputMode === "email" ? ((0, jsx_runtime_1.jsx)(InputSelectionUI_js_1.InputSelectionUI, { type: type, onSelect: (value) => {
                        setData({ emailLogin: value });
                        props.select();
                    }, placeholder: placeholder, name: "email", errorMessage: (input) => {
                        const isValidEmail = (0, validateEmail_js_1.validateEmail)(input.toLowerCase());
                        if (!isValidEmail) {
                            return locale.invalidEmail;
                        }
                        return undefined;
                    }, emptyErrorMessage: emptyErrorMessage, submitButtonText: locale.submitEmail })) : ((0, jsx_runtime_1.jsx)(WalletTypeRowButton_js_1.WalletTypeRowButton, { client: props.client, icon: socialIcons_js_1.emailIcon, onClick: () => {
                        setManualInputMode("email");
                    }, title: locale.emailPlaceholder })) })), isPhoneEnabled && ((0, jsx_runtime_1.jsx)(jsx_runtime_1.Fragment, { children: inputMode === "phone" ? ((0, jsx_runtime_1.jsx)(InputSelectionUI_js_1.InputSelectionUI, { format: "phone", type: type, onSelect: (value) => {
                        // removes white spaces and special characters
                        setData({ phoneLogin: value.replace(/[-\(\) ]/g, "") });
                        props.select();
                    }, placeholder: placeholder, name: "phone", errorMessage: (_input) => {
                        // removes white spaces and special characters
                        const input = _input.replace(/[-\(\) ]/g, "");
                        const isPhone = /^[0-9]+$/.test(input);
                        if (!isPhone && isPhoneEnabled) {
                            return locale.invalidPhone;
                        }
                        return undefined;
                    }, emptyErrorMessage: emptyErrorMessage, submitButtonText: locale.submitEmail })) : ((0, jsx_runtime_1.jsx)(WalletTypeRowButton_js_1.WalletTypeRowButton, { client: props.client, icon: socialIcons_js_1.phoneIcon, onClick: () => {
                        setManualInputMode("phone");
                    }, title: locale.phonePlaceholder })) })), passKeyEnabled && ((0, jsx_runtime_1.jsx)(jsx_runtime_1.Fragment, { children: (0, jsx_runtime_1.jsx)(WalletTypeRowButton_js_1.WalletTypeRowButton, { client: props.client, icon: socialIcons_js_1.passkeyIcon, onClick: () => {
                        handlePassKeyLogin();
                    }, title: locale.passkey }) }))] }));
};
exports.ConnectWalletSocialOptions = ConnectWalletSocialOptions;
const SocialButton = /* @__PURE__ */ (0, styled_1.default)(buttons_js_1.Button)({
    flexGrow: 1,
    "&[data-variant='full']": {
        display: "flex",
        justifyContent: "flex-start",
        padding: index_js_1.spacing.md,
        gap: index_js_1.spacing.sm,
        fontSize: index_js_1.fontSize.md,
        fontWeight: 500,
        transition: "background-color 0.2s ease",
        "&:active": {
            boxShadow: "none",
        },
    },
    "&[data-variant='icon']": {
        padding: index_js_1.spacing.sm,
    },
});
