// thirdweb/dist/cjs/extensions/modular/__generated__/DelayedRevealBatchMetadataERC721/read/encodeBytesUploadMetadata.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isEncodeBytesUploadMetadataSupported = isEncodeBytesUploadMetadataSupported;
exports.encodeEncodeBytesUploadMetadataParams = encodeEncodeBytesUploadMetadataParams;
exports.encodeEncodeBytesUploadMetadata = encodeEncodeBytesUploadMetadata;
exports.decodeEncodeBytesUploadMetadataResult = decodeEncodeBytesUploadMetadataResult;
exports.encodeBytesUploadMetadata = encodeBytesUploadMetadata;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x479eac8a";
const FN_INPUTS = [
    {
        name: "encryptedURI",
        type: "bytes",
        internalType: "bytes",
    },
    {
        name: "provenanceHash",
        type: "bytes32",
        internalType: "bytes32",
    },
];
const FN_OUTPUTS = [
    {
        name: "",
        type: "bytes",
        internalType: "bytes",
    },
];
/**
 * Checks if the `encodeBytesUploadMetadata` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `encodeBytesUploadMetadata` method is supported.
 * @extension MODULAR
 * @example
 * ```ts
 * import { isEncodeBytesUploadMetadataSupported } from "thirdweb/extensions/modular";
 *
 * const supported = await isEncodeBytesUploadMetadataSupported(contract);
 * ```
 */
async function isEncodeBytesUploadMetadataSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "encodeBytesUploadMetadata" function.
 * @param options - The options for the encodeBytesUploadMetadata function.
 * @returns The encoded ABI parameters.
 * @extension MODULAR
 * @example
 * ```ts
 * import { encodeEncodeBytesUploadMetadataParams } "thirdweb/extensions/modular";
 * const result = encodeEncodeBytesUploadMetadataParams({
 *  encryptedURI: ...,
 *  provenanceHash: ...,
 * });
 * ```
 */
function encodeEncodeBytesUploadMetadataParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [
        options.encryptedURI,
        options.provenanceHash,
    ]);
}
/**
 * Encodes the "encodeBytesUploadMetadata" function into a Hex string with its parameters.
 * @param options - The options for the encodeBytesUploadMetadata function.
 * @returns The encoded hexadecimal string.
 * @extension MODULAR
 * @example
 * ```ts
 * import { encodeEncodeBytesUploadMetadata } "thirdweb/extensions/modular";
 * const result = encodeEncodeBytesUploadMetadata({
 *  encryptedURI: ...,
 *  provenanceHash: ...,
 * });
 * ```
 */
function encodeEncodeBytesUploadMetadata(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeEncodeBytesUploadMetadataParams(options).slice(2));
}
/**
 * Decodes the result of the encodeBytesUploadMetadata function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension MODULAR
 * @example
 * ```ts
 * import { decodeEncodeBytesUploadMetadataResult } from "thirdweb/extensions/modular";
 * const result = decodeEncodeBytesUploadMetadataResult("...");
 * ```
 */
function decodeEncodeBytesUploadMetadataResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "encodeBytesUploadMetadata" function on the contract.
 * @param options - The options for the encodeBytesUploadMetadata function.
 * @returns The parsed result of the function call.
 * @extension MODULAR
 * @example
 * ```ts
 * import { encodeBytesUploadMetadata } from "thirdweb/extensions/modular";
 *
 * const result = await encodeBytesUploadMetadata({
 *  contract,
 *  encryptedURI: ...,
 *  provenanceHash: ...,
 * });
 *
 * ```
 */
async function encodeBytesUploadMetadata(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.encryptedURI, options.provenanceHash],
    });
}
