// thirdweb/dist/esm/extensions/modular/__generated__/OpenEditionMetadataERC721/read/onTokenURI.js
import { readContract } from "../../../../../transaction/read-contract.js";
import { encodeAbiParameters } from "../../../../../utils/abi/encodeAbiParameters.js";
import { decodeAbiParameters } from "viem";
import { detectMethod } from "../../../../../utils/bytecode/detectExtension.js";
export const FN_SELECTOR = "0xcfc0cb96";
const FN_INPUTS = [
    {
        name: "_id",
        type: "uint256",
        internalType: "uint256",
    },
];
const FN_OUTPUTS = [
    {
        name: "",
        type: "string",
        internalType: "string",
    },
];
/**
 * Checks if the `onTokenURI` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `onTokenURI` method is supported.
 * @extension MODULAR
 * @example
 * ```ts
 * import { isOnTokenURISupported } from "thirdweb/extensions/modular";
 *
 * const supported = await isOnTokenURISupported(contract);
 * ```
 */
export async function isOnTokenURISupported(contract) {
    return detectMethod({
        contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "onTokenURI" function.
 * @param options - The options for the onTokenURI function.
 * @returns The encoded ABI parameters.
 * @extension MODULAR
 * @example
 * ```ts
 * import { encodeOnTokenURIParams } "thirdweb/extensions/modular";
 * const result = encodeOnTokenURIParams({
 *  id: ...,
 * });
 * ```
 */
export function encodeOnTokenURIParams(options) {
    return encodeAbiParameters(FN_INPUTS, [options.id]);
}
/**
 * Encodes the "onTokenURI" function into a Hex string with its parameters.
 * @param options - The options for the onTokenURI function.
 * @returns The encoded hexadecimal string.
 * @extension MODULAR
 * @example
 * ```ts
 * import { encodeOnTokenURI } "thirdweb/extensions/modular";
 * const result = encodeOnTokenURI({
 *  id: ...,
 * });
 * ```
 */
export function encodeOnTokenURI(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (FN_SELECTOR +
        encodeOnTokenURIParams(options).slice(2));
}
/**
 * Decodes the result of the onTokenURI function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension MODULAR
 * @example
 * ```ts
 * import { decodeOnTokenURIResult } from "thirdweb/extensions/modular";
 * const result = decodeOnTokenURIResult("...");
 * ```
 */
export function decodeOnTokenURIResult(result) {
    return decodeAbiParameters(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "onTokenURI" function on the contract.
 * @param options - The options for the onTokenURI function.
 * @returns The parsed result of the function call.
 * @extension MODULAR
 * @example
 * ```ts
 * import { onTokenURI } from "thirdweb/extensions/modular";
 *
 * const result = await onTokenURI({
 *  contract,
 *  id: ...,
 * });
 *
 * ```
 */
export async function onTokenURI(options) {
    return readContract({
        contract: options.contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.id],
    });
}
