// thirdweb/dist/esm/extensions/lens/__generated__/FollowNFT/read/getFollowerProfileId.js
import { readContract } from "../../../../../transaction/read-contract.js";
import { encodeAbiParameters } from "../../../../../utils/abi/encodeAbiParameters.js";
import { decodeAbiParameters } from "viem";
import { detectMethod } from "../../../../../utils/bytecode/detectExtension.js";
export const FN_SELECTOR = "0x886a65c3";
const FN_INPUTS = [
    {
        type: "uint256",
        name: "followTokenId",
    },
];
const FN_OUTPUTS = [
    {
        type: "uint256",
    },
];
/**
 * Checks if the `getFollowerProfileId` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `getFollowerProfileId` method is supported.
 * @extension LENS
 * @example
 * ```ts
 * import { isGetFollowerProfileIdSupported } from "thirdweb/extensions/lens";
 *
 * const supported = await isGetFollowerProfileIdSupported(contract);
 * ```
 */
export async function isGetFollowerProfileIdSupported(contract) {
    return detectMethod({
        contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "getFollowerProfileId" function.
 * @param options - The options for the getFollowerProfileId function.
 * @returns The encoded ABI parameters.
 * @extension LENS
 * @example
 * ```ts
 * import { encodeGetFollowerProfileIdParams } "thirdweb/extensions/lens";
 * const result = encodeGetFollowerProfileIdParams({
 *  followTokenId: ...,
 * });
 * ```
 */
export function encodeGetFollowerProfileIdParams(options) {
    return encodeAbiParameters(FN_INPUTS, [options.followTokenId]);
}
/**
 * Encodes the "getFollowerProfileId" function into a Hex string with its parameters.
 * @param options - The options for the getFollowerProfileId function.
 * @returns The encoded hexadecimal string.
 * @extension LENS
 * @example
 * ```ts
 * import { encodeGetFollowerProfileId } "thirdweb/extensions/lens";
 * const result = encodeGetFollowerProfileId({
 *  followTokenId: ...,
 * });
 * ```
 */
export function encodeGetFollowerProfileId(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (FN_SELECTOR +
        encodeGetFollowerProfileIdParams(options).slice(2));
}
/**
 * Decodes the result of the getFollowerProfileId function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension LENS
 * @example
 * ```ts
 * import { decodeGetFollowerProfileIdResult } from "thirdweb/extensions/lens";
 * const result = decodeGetFollowerProfileIdResult("...");
 * ```
 */
export function decodeGetFollowerProfileIdResult(result) {
    return decodeAbiParameters(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "getFollowerProfileId" function on the contract.
 * @param options - The options for the getFollowerProfileId function.
 * @returns The parsed result of the function call.
 * @extension LENS
 * @example
 * ```ts
 * import { getFollowerProfileId } from "thirdweb/extensions/lens";
 *
 * const result = await getFollowerProfileId({
 *  contract,
 *  followTokenId: ...,
 * });
 *
 * ```
 */
export async function getFollowerProfileId(options) {
    return readContract({
        contract: options.contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.followTokenId],
    });
}
