// thirdweb/dist/cjs/extensions/lens/__generated__/LensHub/read/getProfile.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isGetProfileSupported = isGetProfileSupported;
exports.encodeGetProfileParams = encodeGetProfileParams;
exports.encodeGetProfile = encodeGetProfile;
exports.decodeGetProfileResult = decodeGetProfileResult;
exports.getProfile = getProfile;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xf08f4f64";
const FN_INPUTS = [
    {
        type: "uint256",
        name: "profileId",
    },
];
const FN_OUTPUTS = [
    {
        type: "tuple",
        components: [
            {
                type: "uint256",
                name: "pubCount",
            },
            {
                type: "address",
                name: "followModule",
            },
            {
                type: "address",
                name: "followNFT",
            },
            {
                type: "string",
                name: "__DEPRECATED__handle",
            },
            {
                type: "string",
                name: "__DEPRECATED__imageURI",
            },
            {
                type: "string",
                name: "__DEPRECATED__followNFTURI",
            },
            {
                type: "string",
                name: "metadataURI",
            },
        ],
    },
];
/**
 * Checks if the `getProfile` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `getProfile` method is supported.
 * @extension LENS
 * @example
 * ```ts
 * import { isGetProfileSupported } from "thirdweb/extensions/lens";
 *
 * const supported = await isGetProfileSupported(contract);
 * ```
 */
async function isGetProfileSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "getProfile" function.
 * @param options - The options for the getProfile function.
 * @returns The encoded ABI parameters.
 * @extension LENS
 * @example
 * ```ts
 * import { encodeGetProfileParams } "thirdweb/extensions/lens";
 * const result = encodeGetProfileParams({
 *  profileId: ...,
 * });
 * ```
 */
function encodeGetProfileParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.profileId]);
}
/**
 * Encodes the "getProfile" function into a Hex string with its parameters.
 * @param options - The options for the getProfile function.
 * @returns The encoded hexadecimal string.
 * @extension LENS
 * @example
 * ```ts
 * import { encodeGetProfile } "thirdweb/extensions/lens";
 * const result = encodeGetProfile({
 *  profileId: ...,
 * });
 * ```
 */
function encodeGetProfile(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeGetProfileParams(options).slice(2));
}
/**
 * Decodes the result of the getProfile function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension LENS
 * @example
 * ```ts
 * import { decodeGetProfileResult } from "thirdweb/extensions/lens";
 * const result = decodeGetProfileResult("...");
 * ```
 */
function decodeGetProfileResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "getProfile" function on the contract.
 * @param options - The options for the getProfile function.
 * @returns The parsed result of the function call.
 * @extension LENS
 * @example
 * ```ts
 * import { getProfile } from "thirdweb/extensions/lens";
 *
 * const result = await getProfile({
 *  contract,
 *  profileId: ...,
 * });
 *
 * ```
 */
async function getProfile(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.profileId],
    });
}
