// thirdweb/dist/cjs/extensions/lens/__generated__/LensHub/read/getProfileIdByHandleHash.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isGetProfileIdByHandleHashSupported = isGetProfileIdByHandleHashSupported;
exports.encodeGetProfileIdByHandleHashParams = encodeGetProfileIdByHandleHashParams;
exports.encodeGetProfileIdByHandleHash = encodeGetProfileIdByHandleHash;
exports.decodeGetProfileIdByHandleHashResult = decodeGetProfileIdByHandleHashResult;
exports.getProfileIdByHandleHash = getProfileIdByHandleHash;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x19e14070";
const FN_INPUTS = [
    {
        type: "bytes32",
        name: "handleHash",
    },
];
const FN_OUTPUTS = [
    {
        type: "uint256",
    },
];
/**
 * Checks if the `getProfileIdByHandleHash` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `getProfileIdByHandleHash` method is supported.
 * @extension LENS
 * @example
 * ```ts
 * import { isGetProfileIdByHandleHashSupported } from "thirdweb/extensions/lens";
 *
 * const supported = await isGetProfileIdByHandleHashSupported(contract);
 * ```
 */
async function isGetProfileIdByHandleHashSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "getProfileIdByHandleHash" function.
 * @param options - The options for the getProfileIdByHandleHash function.
 * @returns The encoded ABI parameters.
 * @extension LENS
 * @example
 * ```ts
 * import { encodeGetProfileIdByHandleHashParams } "thirdweb/extensions/lens";
 * const result = encodeGetProfileIdByHandleHashParams({
 *  handleHash: ...,
 * });
 * ```
 */
function encodeGetProfileIdByHandleHashParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.handleHash]);
}
/**
 * Encodes the "getProfileIdByHandleHash" function into a Hex string with its parameters.
 * @param options - The options for the getProfileIdByHandleHash function.
 * @returns The encoded hexadecimal string.
 * @extension LENS
 * @example
 * ```ts
 * import { encodeGetProfileIdByHandleHash } "thirdweb/extensions/lens";
 * const result = encodeGetProfileIdByHandleHash({
 *  handleHash: ...,
 * });
 * ```
 */
function encodeGetProfileIdByHandleHash(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeGetProfileIdByHandleHashParams(options).slice(2));
}
/**
 * Decodes the result of the getProfileIdByHandleHash function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension LENS
 * @example
 * ```ts
 * import { decodeGetProfileIdByHandleHashResult } from "thirdweb/extensions/lens";
 * const result = decodeGetProfileIdByHandleHashResult("...");
 * ```
 */
function decodeGetProfileIdByHandleHashResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "getProfileIdByHandleHash" function on the contract.
 * @param options - The options for the getProfileIdByHandleHash function.
 * @returns The parsed result of the function call.
 * @extension LENS
 * @example
 * ```ts
 * import { getProfileIdByHandleHash } from "thirdweb/extensions/lens";
 *
 * const result = await getProfileIdByHandleHash({
 *  contract,
 *  handleHash: ...,
 * });
 *
 * ```
 */
async function getProfileIdByHandleHash(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.handleHash],
    });
}
