// thirdweb/dist/esm/extensions/modular/ModularCore/write/installPublishedExtension.js
import { getOrDeployInfraForPublishedContract } from "../../../../contract/deployment/utils/bootstrap.js";
import { installExtension } from "../../__generated__/ModularCore/write/installExtension.js";
/**
 * Install a published extension on a modular contract
 * @param options - The options for installing a published extension
 * @returns A prepared transaction to send
 * @extension MODULAR
 * @example
 * ```ts
 * import { installPublishedExtension } from "thirdweb/extensions/modular";
 * const transaction = installPublishedExtension({
 *  client,
 *  chain,
 *  account,
 *  contract,
 *  extensionName: "MyExtension",
 *  publisherAddress: "0x...",
 * });
 * await sendTransaction({ transaction, account });
 * ```
 */
export function installPublishedExtension(options) {
    const { client, chain, account, contract, extensionName, constructorParams, publisherAddress, extensionData, } = options;
    return installExtension({
        contract,
        asyncParams: async () => {
            const deployedExtension = await getOrDeployInfraForPublishedContract({
                chain,
                client,
                account,
                contractId: extensionName,
                constructorParams: constructorParams || [],
                publisher: publisherAddress,
            });
            return {
                extension: deployedExtension.implementationContract.address,
                data: extensionData || "0x",
            };
        },
    });
}
